File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& mat, int target) {
        if(mat.size() == 0) return false;
        
        int n = mat.size();
        int m = mat[0].size();
        
        int l = 0, h = n*m - 1;
        
        while(l<=h){
            int mid = l + (h-l)/2 ;
            
            if(mat[mid/m][mid%m] == target){
                return true;
            }
            if(mat[mid/m][mid%m] > target){
               
                h = mid-1;
            }
            else{
                  l = mid+1;
            }
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int x) {
        int n = matrix.length;
        int m = matrix[0].length;
        if (matrix.length == 0) {
            return false;
        }

        int low = 0, high = (n * m) - 1;

        while (low <= high) {

            int mid = (low + high) / 2;
            if (matrix[mid / m][mid % m] == x) {
                return true;
            }

            if (matrix[mid / m][mid % m] < x) {

                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return false;
    }
}